### PR TITLE
Add New SSL Native API Test Cases

### PR DESCRIPTION
**Pull request by pepone**
This PR adds four new test cases for SSL native APIs and includes minor test cleanup:

1. **clientValidatesServerUsingValidationCallback**:
   - Verifies that the client can use a validation callback to accept a certificate that would otherwise be rejected.

2. **clientValidatesServerUsingSystemTrustedRootCertificates**:
   - Tests the client's ability to validate `zeroc.com` certificates using default system Certificate Authorities (CAs).

3. **serverValidatesClientUsingValidationCallback**:
   - Client-side counterpart of `clientValidatesServerUsingValidationCallback`, ensuring that the server can use a validation callback to accept client certificates.

4. **serverHotCertificateReload**:
   - Tests the server's ability to reload certificates without restarting.

### CODE REVIEW

File: cpp/src/Ice/SSL/OpenSSLTransceiverI.cpp [hunk 1]
```diff
@@ -36,10 +36,13 @@ extern "C"
 
 namespace
 {
-    std::function<bool(bool, X509_STORE_CTX* ctx, const Ice::SSL::ConnectionInfoPtr&)>
-    createDefaultVerificationCallback()
+    bool defaultVerificationCallback(bool ok, X509_STORE_CTX*, const Ice::SSL::ConnectionInfoPtr&) { return ok; }
```
Comment (pepone) [inline cpp/src/Ice/SSL/OpenSSLTransceiverI.cpp:39]: This avoids creating one callback per connection when the user doesn't provide one.

File: cpp/src/Ice/SSL/OpenSSLTransceiverI.cpp [hunk 2]
```diff
@@ -36,10 +36,13 @@ extern "C"
...
+
+    SSL_CTX* defaultSSLContextSelectionCallback(const string&)
```
Comment (pepone) [inline cpp/src/Ice/SSL/OpenSSLTransceiverI.cpp:41]: We didn't have a default context. This new callback is used when the user doesn't provide one and ensures the default context would use the default platform CAs.

File: cpp/test/IceSSL/configuration/OpenSSLTests.cpp [hunk 3]
```diff
@@ -117,6 +114,84 @@ clientValidatesServerUsingCAFile(Test::TestHelper* helper, const string& testDir
...
+void
+clientValidatesServerUsingSystemTrustedRootCertificates(Test::TestHelper*, const string&)
+{
+    cout << "client validates server using system trusted root certificates... " << flush;
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_client_method());
+    SSL_CTX_set_default_verify_paths(clientSSLContext);
+    Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+        .sslNewSessionCallback = [](::SSL* ssl, const string& host)
+        {
+            // Enable SNI for connecting to the Glacier2 router behind NGINX proxy.
```
Comment (externl) [inline cpp/test/IceSSL/configuration/OpenSSLTests.cpp:178]: SNI is needed here also for CloudFront, which is the first stop and requires SNI. 

I thought we performed SNI by default, is that not the case?

File: cpp/test/IceSSL/configuration/OpenSSLTests.cpp [hunk 4]
```diff
@@ -317,12 +391,76 @@ serverValidatesClientUsingCAFile(Test::TestHelper* helper, const string& testDir
             { SSL_set_verify(ssl, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, nullptr); }};
         Ice::CommunicatorHolder serverCommunicator(createServer(serverAuthenticationOptions, helper));
 
-        Ice::CommunicatorHolder clientCommunicator(createClient(Ice::SSL::ClientAuthenticationOptions{
+        Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
```
Comment (externl) [inline cpp/test/IceSSL/configuration/OpenSSLTests.cpp:394]: what's the change here (for `serverValidatesClientUsingCAFile`) fixing?

File: cpp/test/IceSSL/configuration/SchannelTests.cpp [hunk 5]
```diff
@@ -299,6 +350,7 @@ serverValidatesClientSettingTrustedRootCertificates(Test::TestHelper* helper, co
                 CertDuplicateCertificateContext(serverCertificate);
                 return serverCertificate;
             },
+            .clientCertificateRequired = true,
```
Comment (externl) [inline cpp/test/IceSSL/configuration/SchannelTests.cpp:353]: Should this have been failing before?

File: cpp/test/IceSSL/configuration/SecureTransportTests.cpp [hunk 6]
```diff
@@ -150,9 +155,35 @@ clientRejectsServerSettingTrustedRootCertificates(Test::TestHelper* helper, cons
 }
 
 void
-clientRejectsServerUsingDefaultTrustedRootCertificates(Test::TestHelper* helper, const string& certificatesPath)
+clientValidatesServerUsingSystemTrustedRootCertificates(Test::TestHelper*, const string&)
 {
-    cout << "client rejects server using default trusted root certificates... " << flush;
+    cout << "client validates server using system trusted root certificates... " << flush;
+    Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+        .sslNewSessionCallback = [](SSLContextRef ssl, const string& host)
```
Comment (externl) [inline cpp/test/IceSSL/configuration/SecureTransportTests.cpp:162]: Ditto. Does the already somewhat configured context ref so have SNI set?

File: cpp/test/IceSSL/configuration/SecureTransportTests.cpp [hunk 7]
```diff
@@ -150,9 +155,35 @@ clientRejectsServerSettingTrustedRootCertificates(Test::TestHelper* helper, cons
...
+        {
+            // Enable SNI for connecting to the Glacier2 router behind NGINX proxy.
```
Comment (externl) [inline cpp/test/IceSSL/configuration/SecureTransportTests.cpp:164]: Really for CloudFront

File: cpp/test/IceSSL/configuration/OpenSSLTests.cpp (hunk 3 above)
Comment (pepone) [reply cpp/test/IceSSL/configuration/OpenSSLTests.cpp:178]: > I thought we performed SNI by default, is that not the case?

Not enabled by default, Schannel enables it by default, but with OpenSSL and SecureTransport you have to use their respective APIs.

Keep in mind that with this API we provide mostly not defaults, and neither OpenSSL or SecureTransport enable it by default.

Comment (externl) [reply cpp/test/IceSSL/configuration/OpenSSLTests.cpp:178]: I think this is something we should consider enabling by default.

File: cpp/test/IceSSL/configuration/OpenSSLTests.cpp (hunk 4 above)
Comment (pepone) [reply cpp/test/IceSSL/configuration/OpenSSLTests.cpp:394]: Is still the same test I move the creation of the options out of the method call.

File: cpp/test/IceSSL/configuration/SchannelTests.cpp (hunk 5 above)
Comment (pepone) [reply cpp/test/IceSSL/configuration/SchannelTests.cpp:353]: No. We can improve the test to ensure the client provided a certificate. For a follow-up PR.

File: cpp/test/IceSSL/configuration/SecureTransportTests.cpp (hunk 6 above)
Comment (pepone) [reply cpp/test/IceSSL/configuration/SecureTransportTests.cpp:162]: It doesn't

File: cpp/test/IceSSL/configuration/OpenSSLTests.cpp (hunk 3 above)
Comment (externl) [reply cpp/test/IceSSL/configuration/OpenSSLTests.cpp:178]: We enable SNI by default in 3.7 correct?

Comment (pepone) [reply cpp/test/IceSSL/configuration/OpenSSLTests.cpp:178]: We do not, see https://doc.zeroc.com/ice/3.7/property-reference/icessl#id-.IceSSL.*v3.7-IceSSL.CheckCertName

**Review by externl — APPROVED** [review]

**Review by bernardnormier — APPROVED** [review]
I agree with Joe that we should enable client-side SNI by default (when doable). And remove the IceSSL.CheckCertName property.

**Comment by pepone** [comment]
> I agree with Joe that we should enable client-side SNI by default (when doable). And remove the IceSSL.CheckCertName property.

I think is doable. I will do in a separate PR.

**Pull request merged**